flax/experimental/nnx/nnx/transforms.py

```diff
@@ -541,188 +318,289 @@ def jit(
+    graphdef, diff_state, non_diff_state
   )
-  def grad_fn(*args):
-    nonlocal out_nodes
 
-    _args = list(args)
-    for i, graph_node in diff_graph_nodes.items():
-      diff_state: State = _args[i]
-      graph.update(graph_node, diff_state)
-      _args[i] = graph_node
+  # add nodes to the args
+  for i, arg in diff_graph_nodes.items():
+    _args[i] = arg
 
-    out = f(*_args)
-    out, out_nodes = graph.extract_graph_nodes(out)
+  out = f(*_args)
 
-    _, updates, _ = graph.flatten((input_nodes, out_nodes))
+  out, out_nodes = graph.extract_graph_nodes(out)
 
-    if options.has_aux:
-      loss, aux = out
-      out = (loss, (updates, aux))
-    else:
-      out = (out, updates)
+  graphdef_out, state_out = ctx.split((input_nodes, out_nodes))
+  ctx.clear_idxmap()  # avoid memory leak
 
-    return out
+  if has_aux:
+    loss, aux = out
+    out = (loss, (ctx, graphdef_out, state_out, aux))
+  else:
+    out = (out, (ctx, graphdef_out, state_out))
 
-  out = grad_fn(*_args)
+  return out
 
-  updates: State
-  if options.return_value:
-    if options.has_aux:
-      (loss, (updates, aux)), grads = out
-      out = (loss, aux), grads
-    else:
-      (loss, updates), grads = out
-      out = loss, grads
-  else:
-    if options.has_aux:
-      grads, (updates, aux) = out
-      out = grads, aux
+
+def _grad_general(
+  f: tp.Callable[..., tp.Any],
+  argnums: int | tp.Sequence[int],
+  has_aux: bool,
+  holomorphic: bool,
+  allow_int: bool,
+  reduce_axes: tp.Sequence[AxisName],
+  wrt: filterlib.Filter,
+  return_value: bool,
+) -> tp.Callable[..., tp.Any]:
+  def grad_wrapper(*args):
+    ctx: graph.UpdateContext = graph.UpdateContext()
+    _argnums = _normalize_sequence(argnums)
+    _, input_nodes = graph.extract_graph_nodes(args)
+
+    _args = list(args)
+    diff_graph_nodes: dict[int, tp.Any] = {
+      i: arg
+      for i, arg in enumerate(args)
+      if i in _argnums and graph.is_node(arg)
+    }
+
+    def only_diff(path: tuple, value: tp.Any) -> bool:
+      # diff_graph_nodes is the first element in the tuple
+      return path[0] == 0
+
+    graphdef, diff_state, non_diff_state = ctx.split(
+      (diff_graph_nodes, input_nodes), filterlib.All(wrt, only_diff), ...
+    )  # type: ignore[misc]
+
+    # extract diff_state substates into the args
+    diff_args: list[int] = []
+    if 0 in diff_state:
+      for i, diff_substate in diff_state[0].items():  # type: ignore
+        assert isinstance(i, int)
+        _args[i] = diff_substate
+        diff_args.append(i)
+    transform = jax.value_and_grad if return_value else jax.grad
+
+    _argnums = _argnums[0] if len(_argnums) == 1 else _argnums
+
+    ctx.clear_refmap()  # avoid memory leak
+    out = transform(
+      grad_fn,
+      argnums=_argnums,
+      has_aux=True,
+      holomorphic=holomorphic,
+      allow_int=allow_int,
+      reduce_axes=reduce_axes,
+    )(*_args, f, ctx, graphdef, non_diff_state, has_aux, diff_args)
+
+    updates: State
```

Comment:
Unused?

Reply:
used to force the type annotation on the `update` definitions below

Reply:
I don't see any `updates` definitions below. Which lines are you referring to?